pallets/settlement/src/lib.rs

```diff
@@ -1340,50 +1394,105 @@ impl<T: Trait> Module<T> {
-                                        &leg.from_account_id,
-                                    )
-                                {
+                                if let Ok(from_mercat_pending_state) = Self::get_pending_state(
+                                    &leg.from.did,
+                                    &leg.from_account_id,
+                                    instruction_id,
+                                ) {
                                     if let Ok((to_mercat, to_mercat_balance)) =
                                         Self::get_account_and_balance(
                                             &leg.to.did,
                                             &leg.to_account_id,
                                         )
                                     {
-                                        // Verify the proofs.
-                                        let mut rng = rng::Rng::default();
-                                        let result = TransactionValidator {}.verify_transaction(
-                                            &decoded_justified_tx,
-                                            &from_mercat,
-                                            &from_mercat_balance,
-                                            &to_mercat,
-                                            &to_mercat_balance,
-                                            &[],
-                                            &mut rng,
-                                        );
-                                        if let Ok((new_sender_balance, new_receiver_balance)) =
-                                            result
-                                        {
-                                            <ConfidentialAsset<T>>::update_mercat_account_balance(
-                                                &leg.from.did,
-                                                &leg.from_account_id,
-                                                new_sender_balance,
-                                            );
-                                            <ConfidentialAsset<T>>::update_mercat_account_balance(
-                                                &leg.to.did,
-                                                &leg.to_account_id,
-                                                new_receiver_balance,
-                                            );
+                                        // Get sender's current balance.
+                                        if let Ok((from_mercat, _)) = Self::get_account_and_balance(
+                                            &leg.from.did,
+                                            &leg.from_account_id,
+                                        ) {
+                                            // Verify the proofs.
+                                            let mut rng = rng::Rng::default();
+                                            let result = TransactionValidator {}
```

Comment:
I would rather use `map_err` and `?` here. Otherwise the then-branch is too deep down.

```
let result = TransactionValidator {}
    .verify_transaction( /* ... */ )
    .map_err(|_| (leg_id, FailureReason::GetMercatAccount))?;
```

Reply:
Please also try `TransactionValidator::verify_transaction`. It works if `TransactionValidator` is defined without the annoying `{}`.